whisper/whisper05/whisper.go

```diff
@@ -0,0 +1,583 @@
+	"github.com/ethereum/go-ethereum/rpc"
+
+	"golang.org/x/crypto/pbkdf2"
+
+	"github.com/ethereum/go-ethereum/rlp"
+	set "gopkg.in/fatih/set.v0"
+)
+
+// Whisper represents a dark communication interface through the Ethereum
+// network, using its very own P2P communication layer.
+type Whisper struct {
+	protocol p2p.Protocol
+	filters  *Filters
+
+	privateKeys map[string]*ecdsa.PrivateKey
+	topicKeys   map[string][]byte
+	keyMu       sync.RWMutex
+
+	envelopes   map[common.Hash]*Envelope        // Pool of messages currently tracked by this node
+	messages    map[common.Hash]*ReceivedMessage // Pool of successfully decrypted messages, not expired yet
+	expirations map[uint32]*set.SetNonTS         // Message expiration pool
+	poolMu      sync.RWMutex                     // Mutex to sync the message and expiration pools
+
+	peers  map[*WhisperPeer]struct{} // Set of currently active peers
+	peerMu sync.RWMutex              // Mutex to sync the active peer set
+
+	mailServer *MailServer
+
+	quit chan struct{}
+}
+
+// New creates a Whisper client ready to communicate through the Ethereum P2P network.
+// Param s should be passed if you want to implement mail server, otherwise nil.
+func New(s *MailServer) *Whisper {
+	whisper := &Whisper{
+		privateKeys: make(map[string]*ecdsa.PrivateKey),
+		topicKeys:   make(map[string][]byte),
+		envelopes:   make(map[common.Hash]*Envelope),
+		messages:    make(map[common.Hash]*ReceivedMessage),
+		expirations: make(map[uint32]*set.SetNonTS),
+		peers:       make(map[*WhisperPeer]struct{}),
+		mailServer:  s,
+		quit:        make(chan struct{}),
+	}
+	whisper.filters = NewFilters(whisper)
+
+	// p2p whisper sub protocol handler
+	whisper.protocol = p2p.Protocol{
+		Name:    protocolName,
+		Version: uint(protocolVersion),
+		Length:  2,
+		Run:     whisper.handlePeer,
+	}
+
+	return whisper
+}
+
+// APIs returns the RPC descriptors the Whisper implementation offers
+func (s *Whisper) APIs() []rpc.API {
+	return []rpc.API{
+		{
+			Namespace: protocolName,
+			Version:   protocolVersionStr,
+			Service:   NewPublicWhisperAPI(s),
+			Public:    true,
+		},
+	}
+}
+
+// Protocols returns the whisper sub-protocols ran by this particular client.
+func (self *Whisper) Protocols() []p2p.Protocol {
+	return []p2p.Protocol{self.protocol}
+}
+
+// Version returns the whisper sub-protocols version number.
+func (self *Whisper) Version() uint {
+	return self.protocol.Version
+}
+
+// MarkPeerTrusted marks specific peer trusted, which will allow it
+// to send historic (expired) messages.
+func (self *Whisper) MarkPeerTrusted(peerID *rpc.HexBytes) error {
+	self.peerMu.Lock()
+	defer self.peerMu.Unlock()
+	for p, _ := range self.peers {
+		id := p.peer.ID()
+		if bytes.Equal(*peerID, id[:]) {
+			p.trusted = true
+			return nil
+		}
+	}
+	return fmt.Errorf("Could not find peer with ID: %x", peerID)
+}
+
+func (self *Whisper) RequestHistoricMessages(peerID *rpc.HexBytes, data *rpc.HexBytes) error {
```

Comment:
Please use []byte instead of rpc.HexBytes in Go APIs. They also don't need to be pointers.